solidity/contracts/l2/L2WormholeGateway.sol

```diff
@@ -0,0 +1,275 @@
+///         Wormhole tBTC representation through the bridge in an equal amount.
+/// @dev This contract is supposed to be deployed behind a transparent
+///      upgradeable proxy.
+contract L2WormholeGateway is
+    Initializable,
+    OwnableUpgradeable,
+    ReentrancyGuardUpgradeable
+{
+    using SafeERC20Upgradeable for IERC20Upgradeable;
+
+    /// @notice Reference to the Wormhole Token Bridge contract.
+    IWormholeTokenBridge public bridge;
+
+    /// @notice Wormhole tBTC token representation.
+    IERC20Upgradeable public bridgeToken;
+
+    /// @notice Canonical tBTC token.
+    L2TBTC public tbtc;
+
+    /// @notice Minting limit for this gateway. Useful for early days of testing
+    ///         the system. The gateway can not mint more canonical tBTC than
+    ///         this limit.
+    uint256 public mintingLimit;
+
+    /// @notice The amount of tBTC minted by this contract. tBTC burned by this
+    ///         contract decreases this amount.
+    uint256 public mintedAmount;
+
+    event WormholeTbtcReceived(address receiver, uint256 amount);
+
+    event WormholeTbtcSent(
+        uint256 amount,
+        uint16 recipientChain,
+        bytes32 recipient,
+        uint256 arbiterFee,
+        uint32 nonce
+    );
+
+    event MintingLimitUpdated(uint256 mintingLimit);
+
+    function initialize(
+        IWormholeTokenBridge _bridge,
+        IERC20Upgradeable _bridgeToken,
+        L2TBTC _tbtc
+    ) external initializer {
+        __Ownable_init();
+        __ReentrancyGuard_init();
+
+        require(
+            address(_bridge) != address(0),
+            "Wormhole bridge address must not be 0x0"
+        );
+        require(
+            address(_bridgeToken) != address(0),
+            "Bridge token address must not be 0x0"
+        );
+        require(
+            address(_tbtc) != address(0),
+            "L2TBTC token address must not be 0x0"
+        );
+
+        bridge = _bridge;
+        bridgeToken = _bridgeToken;
+        tbtc = _tbtc;
+        mintingLimit = type(uint256).max;
+    }
+
+    /// @notice This function is called when the user redeems their token on L2.
+    ///         The contract receives Wormhole tBTC representation and mints the
+    ///         canonical tBTC for the user.
+    ///         If the tBTC minting limit has been reached by this contract,
+    ///         instead of minting tBTC the receiver address receives Wormhole
+    ///         tBTC representation.
+    /// @dev Requirements:
+    ///      - The receiver of Wormhole tBTC should be the L2WormholeGateway
+    ///        contract.
+    ///      - The receiver of the canonical tBTC should be abi-encoded in the
+    ///        payload.
+    ///      - The receiver of the canonical tBTC must not be the zero address.
+    /// @param encodedVm A byte array containing a Wormhole VAA signed by the
+    ///        guardians.
+    function receiveWormhole(bytes memory encodedVm) external nonReentrant {
```

Comment:
Slight gas optimization here, you could pass `encodedVM` as `calldata` instead of `memory`.

Reply:
See 51b1ce628e163594971c0e27b7889ac657e6946b, please.